### PR TITLE
webhook: add ability to enforce min memory limit

### PR DESCRIPTION
**Pull request by Redent0r**
Enforce MIN_MEMORY_LIMIT for explicit memory limits on regular and init
containers. No minimum is configured by default, so existing webhook users are not
affected.

Assisted-by: GitHub Copilot
Signed-off-by: Saul Paredes <saulparedes@microsoft.com>

### CODE REVIEW

**Comment by Redent0r** [comment]
> This PR has been opened without activity for 180 days. Please comment on the issue or it will be closed in 7 days.

This is not the PR you are looking to close

File: tools/testing/kata-webhook/main.go [hunk 1]
```diff
@@ -75,6 +76,23 @@ func annotatePodMutator(_ context.Context, ar *kwhmodel.AdmissionReview, obj met
 	kataRuntimeClassName := getRuntimeClass(runtimeClassEnvKey, "kata")
 	pod.Spec.RuntimeClassName = &kataRuntimeClassName
 
+	minMemoryLimit, foundMinMemoryLimit := os.LookupEnv("MIN_MEMORY_LIMIT")
+
+	if foundMinMemoryLimit {
+		minMemoryLimitVal := resource.MustParse(minMemoryLimit)
+		for i := range pod.Spec.Containers {
+			if pod.Spec.Containers[i].Resources.Limits == nil {
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/main.go]: enforcing for initContainers. Looks like we can't set limits on ephemeral containers https://k8s.info/docs/expert/ephemeral-containers#1-what-are-ephemeral-containers  (nor they are defined inside Pod.Spec)

File: tools/testing/kata-webhook/main_test.go [hunk 2]
```diff
@@ -58,3 +59,86 @@ func TestAnnotatePodMutator(t *testing.T) {
...
+			container: corev1.Container{
+				Resources: corev1.ResourceRequirements{Limits: corev1.ResourceList{
+					corev1.ResourceCPU: resource.MustParse("1"),
+				}},
+			},
+		},
+		{
+			name:      "container without any resources is left alone",
+			container: corev1.Container{},
+		}
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: fixed

File: tools/testing/kata-webhook/README.md [hunk 3]
```diff
@@ -58,3 +58,24 @@ those with
...
+## Enforcing a minimum memory limit
+
+Kata pods run inside a VM, so a memory limit that is comfortable under `runc`
+can be too small once the guest kernel and agent are accounted for. The webhook
+can raise any explicit memory limit that falls below a configured floor.
+
+No minimum is enforced by default. To enable it, set the `min_memory_limit` key
+of the `kata-webhook` ConfigMap to a
+[quantity](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/),
+either by editing the [webhook file](deploy/webhook.yaml)
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/README.md]: fixed

File: tools/testing/kata-webhook/main_test.go [hunk 4]
```diff
@@ -58,3 +59,86 @@ func TestAnnotatePodMutator(t *testing.T) {
 		})
 	}
 }
+
+func TestEnforceMinMemoryLimit(t *testing.T) {
+	const minimum = "256Mi"
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go:154]: added

File: tools/testing/kata-webhook/README.md [hunk 5]
```diff
@@ -58,3 +58,24 @@ those with
...
+## Enforcing a minimum memory limit
+
+Kata pods run inside a VM, so a memory limit that is comfortable under `runc`
+can be too small once the guest kernel and agent are accounted for. The webhook
+can raise any explicit memory limit that falls below a configured floor.
+
+No minimum is enforced by default. To enable it, set the `min_memory_limit` key
+of the `kata-webhook` ConfigMap to a
+[quantity](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/)
+by editing the [webhook file](deploy/webhook.yaml)
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/README.md]: updated to specify we need to redeploy

File: tools/testing/kata-webhook/main.go [hunk 6]
```diff
@@ -30,6 +31,42 @@ func getRuntimeClass(runtimeClassKey, defaultRuntimeClass string) string {
 	return defaultRuntimeClass
 }
 
+// minMemoryLimit is the enforced floor, or nil when enforcement is disabled.
+var minMemoryLimit *resource.Quantity
+
+// An invalid value disables enforcement rather than taking the webhook down,
+// since failurePolicy is Ignore and pods would then bypass the runtime class too.
```
Comment (sprt) [inline tools/testing/kata-webhook/main.go]: I don't think this makes sense to me. The webhook should return errors, it's Kubelet that enforces failurePolicy.

So sounds like we should error out when we can't parse a found limit.

File: tools/testing/kata-webhook/main.go [hunk 7]
```diff
@@ -30,6 +31,42 @@ func getRuntimeClass(runtimeClassKey, defaultRuntimeClass string) string {
...
+func configureMinMemoryLimit() *resource.Quantity {
+	value, found := os.LookupEnv("MIN_MEMORY_LIMIT")
```
Comment (sprt) [inline tools/testing/kata-webhook/main.go]: We shouldn't introduce a global. Instead let's turn this function into `getMinMemoryLimit()`, similar to `getRuntimeClass()`.

File: tools/testing/kata-webhook/README.md [hunk 8]
```diff
@@ -58,3 +58,34 @@ those with
...
+> `MIN_MEMORY_LIMIT` is a ConfigMap-backed environment variable, so it is only
+> resolved when the webhook pod starts. Set it before running
+> `kubectl apply -f deploy/`, or on an existing installation, apply the updated
+> manifest and then restart the Deployment:
+>
+> ```bash
+> kubectl apply -f deploy/webhook.yaml
+> kubectl rollout restart deployment/pod-annotate-webhook
+> ```
+
```
Comment (sprt) [inline tools/testing/kata-webhook/README.md]: I would remove this, this is not specific to this config and we're not teaching K8s anyway.

File: tools/testing/kata-webhook/main_test.go [hunk 9]
```diff
@@ -58,3 +60,139 @@ func TestAnnotatePodMutator(t *testing.T) {
...
+	if want == "" {
+		if ok {
+			t.Errorf("expected no memory limit, got %s", got.String())
+		}
+		return
+	}
+	if !ok {
+		t.Fatalf("expected memory limit %s, got none", want)
+	}
+	if got.String() != want {
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: * Structuing the code so we don't need returns
* Not using `t.Fatalf()` as that would stop the caller loop
```suggestion
	t.Helper()
	
	got, ok := limits[corev1.ResourceMemory]
    if ok {
        if want == "" {
            t.Errorf("expected no memory limit, got %s", got.String())
        } else if got.String() != want {
		    t.Errorf("expected memory limit %s, got %s", want, got.String())
        }
    } else {
		t.Fatalf("expected memory limit %s, got none", want)
	}
```

File: tools/testing/kata-webhook/main.go (hunk 6 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main.go]: makes sense, the user can change the failurePolicy to always if they want to. The webhook returns an error regardless

File: tools/testing/kata-webhook/main.go (hunk 7 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main.go]: I thought I saw a copilot feedback (or maybe my own agent) about having to reparse the memory limit on each request. This was trying to address that, but I can't find it now. But on second look, adding a global to avoid reparsing may be worst, and parsing is not expensive

File: tools/testing/kata-webhook/README.md [hunk 10]
```diff
@@ -58,3 +58,22 @@ those with
...
+## Enforcing a minimum memory limit
+
+Kata pods run inside a VM, so a memory limit that is comfortable under `runc`
+can be too small once the guest kernel and agent are accounted for. The webhook
+can raise any explicit memory limit that falls below a configured floor.
+
+No minimum is enforced by default. To enable it, set the `min_memory_limit` key
+of the `kata-webhook` ConfigMap in the [webhook file](deploy/webhook.yaml) to a
+[quantity](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/).
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/README.md:70]: no need to teach k8s https://github.com/kata-containers/kata-containers/pull/12652#discussion_r3975614542

File: tools/testing/kata-webhook/main.go [hunk 11]
```diff
@@ -68,6 +98,16 @@ func annotatePodMutator(_ context.Context, ar *kwhmodel.AdmissionReview, obj met
 		return &kwhmutating.MutatorResult{}, nil
 	}
 
+	minMemoryLimitEnvKey := "MIN_MEMORY_LIMIT"
+	minMemoryLimit, err := getMinMemoryLimit(minMemoryLimitEnvKey)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/main.go:105]: we want to return an error instead of silently ignoring. If a user does not want to allow pods that are not mutated correctly, they can set `failurePolicy: always` https://github.com/kata-containers/kata-containers/pull/12652#discussion_r3982615682

[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/main.go:105]: as intended, users can set failurePolicy: always if they want to

[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/main.go:105]: "rather than merely logging and skipping memory enforcement as described" ah, you meant the commit message I think. Amended

File: tools/testing/kata-webhook/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/README.md:70]: No k8s teaching here

File: tools/testing/kata-webhook/main.go [hunk 12]
```diff
@@ -68,6 +98,15 @@ func annotatePodMutator(_ context.Context, ar *kwhmodel.AdmissionReview, obj met
 		return &kwhmutating.MutatorResult{}, nil
 	}
 
+	minMemoryLimit, err := getMinMemoryLimit("MIN_MEMORY_LIMIT")
```
[the remark this answers is not in the document]
Comment (Redent0r) [reply tools/testing/kata-webhook/main.go]: oh you are saying that we should check ar.Namespace == "sonobuoy", instead of pod.GetNamespace() == "sonobuoy" because of existing comment 

// The Namespace is not always available in the pod Spec
// specially when operators create the pods. Hence access
// the Namespace in the actual request (vs the object)
	
I'd say valid but let's address on separate PR if needed. The code change in this PR is not related to namesapace validation

File: tools/testing/kata-webhook/main_test.go
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go:37]: Let's remove the new Test functions and instead cover those cases here.

Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go:37]: updated. I thought it would end up a big convoluted test case but its not

File: tools/testing/kata-webhook/main_test.go [hunk 13]
```diff
@@ -6,55 +6,226 @@ package main
...
-		wantMutated  bool
+		name          string
+		nsBlacklist   map[string]bool
+		nsOnlyRegexp  *regexp.Regexp
+		minMemory     string
+		setMinMemory  bool
+		wantMutated   bool
+		wantErr       bool
+		wantLimit     string
+		wantInitLimit string
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: Both final limits should be the same, no? Then we only need one `wantLimit`

File: tools/testing/kata-webhook/main_test.go
Comment (sprt) [reply tools/testing/kata-webhook/main_test.go:31]: You forgot TestEnforceMinMemoryLimit. Retain all the test cases but put them in the existing Test.

File: tools/testing/kata-webhook/main_test.go (hunk 13 above)
Comment (sprt) [reply tools/testing/kata-webhook/main_test.go]: Ah no not if existing limits < max limit, please ignore.

File: tools/testing/kata-webhook/main_test.go [hunk 14]
```diff
@@ -6,55 +6,231 @@ package main
...
 
 	kwhmodel "github.com/slok/kubewebhook/v2/pkg/model"
 )
 
 func TestAnnotatePodMutator(t *testing.T) {
+	const (
+		initialLimit     = "128Mi"
+		initialInitLimit = "64Mi"
+		enforcedLimit    = "256Mi"
+	)
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: Let's actually remove these constants and inline them, they make the test cases below harder to read.

File: tools/testing/kata-webhook/main_test.go [hunk 15]
```diff
@@ -6,55 +6,231 @@ package main
...
+			if !tt.setMinMemory {
+				if err := os.Unsetenv(minMemoryLimitEnvKey); err != nil {
+					t.Fatalf("failed to unset %s: %v", minMemoryLimitEnvKey, err)
+				}
+			}
+
+			container := tt.container
+			if container == nil {
+				container = memLimitContainer(initialLimit)
+			}
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: We should remove this code and force tt.container to be a container (make it a struct instead of a pointer to a struct) - current test cases are hard to reason about.

File: tools/testing/kata-webhook/main_test.go [hunk 16]
```diff
@@ -6,55 +6,231 @@ package main
...
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			whPolicy = &policy{nsBlacklist: tt.nsBlacklist, nsOnlyRegexp: tt.nsOnlyRegexp}
 
+			t.Setenv(minMemoryLimitEnvKey, tt.minMemory)
+			if !tt.setMinMemory {
+				if err := os.Unsetenv(minMemoryLimitEnvKey); err != nil {
+					t.Fatalf("failed to unset %s: %v", minMemoryLimitEnvKey, err)
+				}
+			}
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: Let's:

* Remove the setMinMemory field
* Make minMemory a *string
* Only call Setenv if minMemory != nil
* Remove the Unsetenv() call

File: tools/testing/kata-webhook/main_test.go [hunk 17]
```diff
@@ -6,55 +6,231 @@ package main
...
+				container = memLimitContainer(initialLimit)
+			}
+
 			pod := &corev1.Pod{
 				ObjectMeta: metav1.ObjectMeta{
 					Namespace: "testing-namespace",
 				},
+				Spec: corev1.PodSpec{
+					Containers:     []corev1.Container{*container},
+					InitContainers: []corev1.Container{*memLimitContainer(initialInitLimit)},
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: We should pass `container` here, no?

File: tools/testing/kata-webhook/main_test.go [hunk 18]
```diff
@@ -6,55 +6,231 @@ package main
...
+			name:          "no filters",
+			wantMutated:   true,
+			wantLimit:     initialLimit,
+			wantInitLimit: initialInitLimit,
+		},
+		{
+			name:          "matching nsBlacklist",
+			nsBlacklist:   map[string]bool{"testing-namespace": true},
+			wantLimit:     initialLimit,
+			wantInitLimit: initialInitLimit,
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: Specify wantMutated explicitely as before, here and in `nonmatching nsOnlyRegexp`

File: tools/testing/kata-webhook/main_test.go (hunk 17 above)
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: This is confusing, let's introduce a new initContainer field into the test struct.

File: tools/testing/kata-webhook/main_test.go [hunk 19]
```diff
@@ -6,55 +6,231 @@ package main
...
+			minMemory:    enforcedLimit,
+			setMinMemory: true,
+			container: &corev1.Container{
+				Resources: corev1.ResourceRequirements{Limits: corev1.ResourceList{
+					corev1.ResourceCPU: resource.MustParse("1"),
+				}},
+			},
+			wantMutated:   true,
+			wantInitLimit: enforcedLimit,
+		},
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: To remove, covered by the next test

File: tools/testing/kata-webhook/main_test.go [hunk 20]
```diff
@@ -6,55 +6,231 @@ package main
...
+		{
+			name:          "container without any resources is left alone",
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: "left alone" but wantMutated is true?

**Review by sprt — COMMENTED** [review]
Test code feels kinda hacky, please carefully review AI code

File: tools/testing/kata-webhook/main_test.go (hunk 16 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: I think we need unsetEnv() to remove the env from a previous test case. So let's say:
```
case1: minMemory: 256
case2: minMemory: nil

for case in cases:
  if minMemory:
    setEnv(minMemory)
  ... // at the end after asserting
  // if we don't call unsetEnv, case 2 still gets the minMemory from case 1
  // unsetEnv(minMemoryLimitEnvKey)
```

Maybe the form above is more readable, as we don't need the setMinMemory field. But we still need to call unsetEnv  I think

File: tools/testing/kata-webhook/main_test.go (hunk 20 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: wantMutated wants the runtimeClass to be mutated, which we still expect in the scenarios where the limit is not. I agree it doesn't read well with "left alone", I'll improve the names

File: tools/testing/kata-webhook/main_test.go (hunk 14 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: staged that here to see what it looks like before squashing https://github.com/kata-containers/kata-containers/pull/12652/changes/8cc077ba98af3b3d7e61f9e430fa4c4a4a1bdade (I think it's more readable)

File: tools/testing/kata-webhook/main_test.go (hunk 20 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: improved

File: tools/testing/kata-webhook/main_test.go (hunk 15 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: addressed in https://github.com/kata-containers/kata-containers/pull/12652/changes/8cc077ba98af3b3d7e61f9e430fa4c4a4a1bdade . Something I also did there that makes it more readable IMO is only specify container/initContainer if it's relevant to the test case

File: tools/testing/kata-webhook/main_test.go (hunk 17 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: addressed in https://github.com/kata-containers/kata-containers/pull/12652/changes/8cc077ba98af3b3d7e61f9e430fa4c4a4a1bdade

File: tools/testing/kata-webhook/main_test.go (hunk 16 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: Actually
```
for case in cases:
  if minMemory:
    setEnv(minMemory)
  ... // at the end after asserting
  unsetEnv(minMemoryLimitEnvKey)
```

is not great either because MIN_MEMORY_LIMIT may be set in the dev environment on the first loop iteration. So maybe 
```
for case in cases:
  if minMemory != nil:
    setEnv(minMemory)
  else:
    unsetEnv(minMemoryLimitEnvKey)
  ... // assert
```

Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: trying to address this in https://github.com/kata-containers/kata-containers/pull/12652/changes/fa6b6194c21ce34b980faa83489a29edba215918

**Comment by Redent0r** [comment]
> ### 🔵 Needs a closer look
> Invalid configuration behavior contradicts the PR description.
> 
> Review details

you are right it contradicts PR description. Updated PR description

File: tools/testing/kata-webhook/main_test.go (hunk 16 above)
Comment (sprt) [reply tools/testing/kata-webhook/main_test.go]: Setenv cleans up automatically https://pkg.go.dev/testing#T.Setenv

Your manual UnsetEnv is only useful if we assume that the env var could be set outside the test, which it shouldn't anyway

File: tools/testing/kata-webhook/main_test.go (hunk 20 above)
Comment (sprt) [reply tools/testing/kata-webhook/main_test.go]: Ahh gotcha, let's rename that var to `wantMutatedRuntimeClass` then

File: tools/testing/kata-webhook/main_test.go [hunk 21]
```diff
@@ -6,55 +6,195 @@ package main
...
 			result, err := annotatePodMutator(context.Background(), ar, pod)
-			if err != nil {
-				t.Fatalf("unexpected error: %v", err)
+			if tt.wantErr {
+				if err == nil {
+					t.Errorf("expected an error for an invalid %s", minMemoryLimitEnvKey)
+				}
+			} else if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			} else {
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go]: nit: simpler handling like for wantMutated 
```suggestion
			if err != tt.wantErr {
				t.Errorf("expected error: %v, got: %v", wantErr, err)
			} else if err != nil {
```

File: tools/testing/kata-webhook/main_test.go [hunk 22]
```diff
@@ -6,55 +6,195 @@ package main
...
+		{
+			name:         "matching nsOnlyRegexp",
+			nsOnlyRegexp: regexp.MustCompile("^testing-.*$"),
+			wantMutated:  true,
+		},
+		{
+			name:         "nonmatching nsOnlyRegexp",
+			nsOnlyRegexp: regexp.MustCompile(".*nonexisting.*"),
+			wantMutated:  false,
+		},
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go:55]: Please keep specifying all the fields that were specified before

File: tools/testing/kata-webhook/main_test.go [hunk 23]
```diff
@@ -6,55 +6,195 @@ package main
...
+		{
+			name:      "zero min memory",
+			minMemory: ptr("0"),
+			wantErr:   true,
+		},
+		{
+			name:      "negative min memory",
+			minMemory: ptr("-1Mi"),
+			wantErr:   true,
+		},
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go:127]: nit: let's add "returns error" to name

File: tools/testing/kata-webhook/main_test.go [hunk 24]
```diff
@@ -6,55 +6,195 @@ package main
...
+			wantMutated: true,
+			wantLimit:   "268435456",
+		},
+		{
+			name:      "container without a memory limit stays unlimited",
+			minMemory: ptr("256Mi"),
+			container: corev1.Container{
+				Resources: corev1.ResourceRequirements{Limits: corev1.ResourceList{
+					corev1.ResourceCPU: resource.MustParse("1"),
+				}},
```
Comment (sprt) [inline tools/testing/kata-webhook/main_test.go:103]: nit: remove resources to match test name and add explicit wantLimit

File: tools/testing/kata-webhook/main_test.go (hunk 16 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go]: ahh I didn't know t.setEnv did clean up. Fair enough

File: tools/testing/kata-webhook/main_test.go (hunk 24 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go:103]: Adding explicit wantLimit but also I'd argue to leave the rest as is:
- Right now it matches the test name "container without a memory limit stays unlimited" - resources are only specifying a CPU limit
- we are testing that specifically unset memory limits keep the same memory limit: currentMemoryLimit, ok := **containers[i].Resources.Limits[corev1.ResourceMemory]** in https://github.com/microsoft/kata-containers/blob/fa6b6194c21ce34b980faa83489a29edba215918/tools/testing/kata-webhook/main.go#L55

If we remove resources , we should update the name to  "container without resources keep memory limit unlimited". But I think that coverage is looser than what we have currently. We would only be testing that unset resources keep the same memory limit: currentMemoryLimit, ok := **containers[i].Resources**.Limits[corev1.ResourceMemory]

Comment (sprt) [reply tools/testing/kata-webhook/main_test.go:103]: Up to you, but IMO we're arbitrarily cherry-picking an edge case, that's why in my previous comment  https://github.com/kata-containers/kata-containers/pull/12652#discussion_r3995726708 I was asking to remove this test and keep the other one.

**Review by sprt — APPROVED** [review]
LGTM - would be nice to add this test to the CI in a follow-up PR.

**Comment by Redent0r** [comment]
pushed squashed changes

**Review by manuelh-dev — APPROVED** [review]

File: tools/testing/kata-webhook/main_test.go (hunk 24 above)
Comment (Redent0r) [reply tools/testing/kata-webhook/main_test.go:103]: Leaving as is for now. The omitted initContainer (initialized to corev1.Container{}) and omitted wantInitLimit (initialized to "") in the same test case already validate that a container with no resources doesn't get a memory limit either, so both are covered here.
 
I'd also keep the CPU limit because documenting that
 ```
spec:
  containers:
    - name: app
      resources:
        limits:
          cpu: "1"
 ```
leaves the memory limit alone is slightly more informative than the empty case: it shows that some limit can be set as long as it isn't memory. This reminds me of [runtime-rs: size static sandboxes per resource by Redent0r · Pull Request #13787 · kata-containers/…](https://github.com/kata-containers/kata-containers/pull/13787) , where is not always clear if code  behaviour is resource independent or not
